core/src/query/store.rs

```diff
@@ -0,0 +1,348 @@
+                    }
+                    to_remove = remove_receiver.recv() => {
+                        let Some(query_id) = to_remove else {
+                            iroha_logger::info!(ALL_HANDLERS_DROPPED);
+                            break;
+                        };
+                        let live_query_opt = self.remove(&query_id);
+                        let _ = remove_result_sender.send(live_query_opt).await;
+                    }
+                    _ = self.notify_shutdown.notified() => {
+                        iroha_logger::info!("Query cleanup service is being shut down.");
+                        break;
+                    }
+                    else => break,
+                }
+                tokio::task::yield_now().await;
+            }
+        });
+
+        LiveQueryStoreHandle {
+            insert_sender,
+            remove_channel: Arc::new(Mutex::new(RemoveChannel(
+                remove_sender,
+                remove_result_receiver,
+            ))),
+        }
+    }
+
+    fn insert(&self, query_id: String, live_query: LiveQuery) {
+        self.queries.insert(query_id, (live_query, Instant::now()));
+    }
+
+    fn remove(&self, query_id: &str) -> Option<LiveQuery> {
+        self.queries.remove(query_id).map(|(_, (output, _))| output)
+    }
+}
+
+/// Handle to interact with [`LiveQueryStore`].
+#[derive(Clone)]
+pub struct LiveQueryStoreHandle {
+    insert_sender: mpsc::Sender<(String, LiveQuery)>,
+    remove_channel: Arc<Mutex<RemoveChannel>>,
+}
+
+struct RemoveChannel(mpsc::Sender<String>, mpsc::Receiver<Option<LiveQuery>>);
+
+impl LiveQueryStoreHandle {
+    /// Validate and execute query performing sorting and pagination.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::ConnectionClosed`] if [`QueryService`] is dropped,
+    /// - Otherwise throws up query handling errors.
+    pub fn handle_query<Q>(
+        &self,
+        query_with_params: WithQueryParameters<Q>,
+        wsv: &WorldStateView,
+    ) -> Result<BatchedResponse<Value>>
+    where
+        Q: ValidateQuery,
+    {
+        let WithQueryParameters {
+            query,
+            fetch_size,
+            sorting,
+            pagination,
+            cursor,
+        } = query_with_params;
+
+        let query: QueryBox = query.validate_query(wsv)?;
+
+        self.handle_query_unchecked(
+            WithQueryParameters {
+                query,
+                fetch_size,
+                sorting,
+                pagination,
+                cursor,
+            },
+            wsv,
+        )
+    }
+
+    /// Execute query performing sorting and pagination without any validation.
+    ///
+    /// Usable for queries executed by validator.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::ConnectionClosed`] if [`QueryService`] is dropped,
+    /// - Otherwise throws up query handling errors.
+    pub fn handle_query_unchecked(
```

Comment:
what is `unchecked`? You say that it doesn't validate query? Do you mean it doesn't go thgrough executor validation?

Reply:
Yes, and other checks like signature check as well, because it's just a `QueryBox` not `SignedQuery`

Reply:
it just seems to me like you are abusing the suffix `_unchecked` here

Reply:
Haha, maybe. Ok, anyway, after your suggestion [here](https://github.com/hyperledger/iroha/pull/3958#discussion_r1347000475) I will move away validation and execution from handle